src/cluster/index.js

```diff
@@ -163,14 +163,23 @@ module.exports = class Cluster {
    */
   async addMultipleTargetTopics(topics) {
     const previousSize = this.targetTopics.size
+    const previousTopics = new Set(this.targetTopics)
     for (const topic of topics) {
       this.targetTopics.add(topic)
     }
 
     const hasChanged = previousSize !== this.targetTopics.size || !this.brokerPool.metadata
 
     if (hasChanged) {
-      await this.refreshMetadata()
+      try {
+        await this.refreshMetadata()
+      } catch (e) {
+        if (e.type === 'INVALID_TOPIC_EXCEPTION') {
+          this.targetTopics = previousTopics
```

Comment:
By storing the previous topics per method call, is there a possibility of us introducing a race condition here? I'm too unfamiliar with this part of the codebase whether that's something reasonable to expect or not.

If we _are_ creating a race condition here, the most simple way to mitigate in my mind is to store the previous topics on the cluster instance and use that (or a Promise of it) as a lock to only allow a single inflight call. Alternatively, we could be more careful with adding new topics to the targets list, only revert the ones for which we got the exception (which should be idempotent, so deletion of them of a Set should be without consequence).

Reply:
Ah, you are a learned and erudite gentleman. I didn't think about the possibility of a race condition here, but you are indeed correct that that could happen.

Long-term I think we should actually remove the list of target topics, if at all possible, as mentioned in #651, but in the meantime I think a simple lock here would be fine.

Reply:
Fixed in `091f5c`.

Reply:
You reckon it's worth adding a test case that confirm the lock works as expected? It could be useful, but triggering race conditions reliably is hard and an ingredient for flaky tests.

Reply:
That was exactly my thinking. I would like to test it, but it also sounds nightmareish to get such a test to work consistently.